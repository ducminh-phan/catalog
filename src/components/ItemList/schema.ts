import { makeRequired, makeValidate } from "mui-rff";
import * as yup from "yup";

const schema = yup.object().shape({
  name: yup.string().required().min(5),
  description: yup.string().required(),
  price: yup.number().required(),
});

export const validate = makeValidate(schema);

export const required = makeRequired(schema);

export type FormData = yup.InferType<typeof schema>;
