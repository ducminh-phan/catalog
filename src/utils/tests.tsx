import { StylesProvider } from "@material-ui/styles";
import { render as rtlRender, RenderResult } from "@testing-library/react";
import { Rule, StyleSheet } from "jss";
import React, { ComponentType, ReactElement, ReactNode } from "react";
import { Provider } from "react-redux";

import Notification from "components/Notification";
import { AuthProvider } from "contexts/auth";
import { NotificationProvider } from "contexts/notification";
import store from "store";

import Request from "./api";
import { UserResponse } from "./auth";
import storage from "./storage";

// Custom logic to make Material UI class names consistent in snapshots
const generateClassName = (
  rule: Rule,
  styleSheet?: StyleSheet<string>,
): string => `${styleSheet?.options.classNamePrefix}-${rule.key}`;

const render = (ui: ReactElement, renderOptions = {}): RenderResult => {
  const Wrapper = ({ children }: { children: ReactNode }): ReactElement => {
    return (
      <StylesProvider generateClassName={generateClassName}>
        <Provider store={store}>
          <NotificationProvider>
            <AuthProvider>{children}</AuthProvider>
            <Notification />
          </NotificationProvider>
        </Provider>
      </StylesProvider>
    );
  };

  return rtlRender(ui, { wrapper: Wrapper as ComponentType, ...renderOptions });
};

export const setupUser = (): void => {
  storage.setToken("accessToken");
  expect(storage.getToken()).toBe("accessToken");
  jest.spyOn(Request, "get").mockImplementation(
    (): Promise<UserResponse> =>
      Promise.resolve({
        email: "minhπ",
        username: "minhπ",
        created: "",
        updated: "",
        name: "minhπ",
        id: 0,
      }),
  );
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { render };
