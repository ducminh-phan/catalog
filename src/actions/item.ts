import { createAsyncThunk } from "@reduxjs/toolkit";

import Request from "utils/api";
import * as types from "utils/types";

type FetchItemsRequest = types.PaginationParams & {
  categoryId?: string;
};

export type FetchItemsResponse = {
  totalItems: number;
  items: types.Item[];
};

export const fetchItems = createAsyncThunk(
  "items/fetch",
  ({ categoryId, offset, limit }: FetchItemsRequest) =>
    Request.get<FetchItemsResponse>(
      `categories/${categoryId}/items?offset=${offset}&limit=${limit}`,
    ),
);

type EditItemsRequest = {
  categoryId?: string;
  itemId: string;
  data: {
    name: string;
    description: string;
    price: number;
  };
};

type EditItemsResponse = types.Item;

export const editItem = createAsyncThunk(
  "items/edit",
  ({ categoryId, itemId, data }: EditItemsRequest) =>
    Request.put<EditItemsResponse>(
      `categories/${categoryId}/items/${itemId}`,
      data,
    ),
);

type AddItemRequest = {
  categoryId?: string;
  data: {
    name: string;
    description: string;
    price: number;
  };
};

type AddItemResponse = types.Item;

export const addItem = createAsyncThunk(
  "item/add",
  ({ categoryId, data }: AddItemRequest) =>
    Request.post<AddItemResponse>(`categories/${categoryId}/items`, data),
);

type DeleteItemRequest = {
  categoryId?: string;
  itemId: string;
};

export const deleteItem = createAsyncThunk(
  "item/delete",
  ({ categoryId, itemId }: DeleteItemRequest) =>
    Request.delete(`categories/${categoryId}/items/${itemId}`),
);
