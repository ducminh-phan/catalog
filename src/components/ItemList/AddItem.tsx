import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
} from "@material-ui/core";
import AddIcon from "@material-ui/icons/Add";
import { unwrapResult } from "@reduxjs/toolkit";
import { TextField } from "mui-rff";
import React, { ReactNode } from "react";
import { Form } from "react-final-form";
import { connect, ConnectedProps } from "react-redux";

import * as itemActions from "actions/item";
import { useNotification } from "contexts/notification";

import { FormData, required, validate } from "./schema";

const mapDispatchToProps = {
  addItem: itemActions.addItem,
};

const connector = connect(null, mapDispatchToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;

type Props = PropsFromRedux & {
  categoryId?: string;
};

const AddItem = (props: Props): React.ReactElement => {
  const { categoryId, addItem } = props;
  const [open, setOpen] = React.useState(false);
  const { setNotification } = useNotification();

  const handleOpen = (): void => {
    setOpen(true);
  };

  const handleClose = (): void => {
    setOpen(false);
  };

  const handleAdd = async (data: FormData): Promise<void> => {
    const resultAction = await addItem({
      categoryId,
      data,
    });

    handleClose();

    try {
      unwrapResult(resultAction);
      setNotification({
        severity: "success",
        message: "Add item successfully",
      });
    } catch (e) {
      setNotification({
        severity: "error",
        message: e.message ?? "Add item failed",
      });
    }
  };

  return (
    <>
      <Button color="inherit" onClick={handleOpen}>
        <AddIcon />
      </Button>
      <Dialog
        open={open}
        onClose={handleClose}
        aria-labelledby="item-add-form-title"
        fullWidth
      >
        <DialogTitle id="item-add-form-title">Add Item</DialogTitle>
        <DialogContent>
          <Form
            onSubmit={handleAdd}
            validate={validate}
            render={({ handleSubmit, invalid }): ReactNode => (
              <form onSubmit={handleSubmit} noValidate>
                <TextField label="Name" name="name" required={required.name} />
                <TextField
                  label="Description"
                  name="description"
                  required={required.description}
                />
                <TextField
                  label="Price"
                  name="price"
                  type="number"
                  required={required.price}
                />
                <DialogActions>
                  <Button type="button" onClick={handleClose}>
                    Cancel
                  </Button>
                  <Button color="primary" type="submit" disabled={invalid}>
                    Submit
                  </Button>
                </DialogActions>
              </form>
            )}
          />
        </DialogContent>
      </Dialog>
    </>
  );
};

export default connector(AddItem);
