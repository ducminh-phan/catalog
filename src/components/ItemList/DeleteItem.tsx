import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle,
} from "@material-ui/core";
import DeleteIcon from "@material-ui/icons/Delete";
import { unwrapResult } from "@reduxjs/toolkit";
import React from "react";
import { connect, ConnectedProps } from "react-redux";

import * as itemActions from "actions/item";
import { useNotification } from "contexts/notification";

const mapDispatchToProps = {
  deleteItem: itemActions.deleteItem,
};

const connector = connect(null, mapDispatchToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;

type Props = PropsFromRedux & {
  categoryId?: string;
  itemId: string;
};

const DeleteItem = (props: Props): React.ReactElement => {
  const { categoryId, itemId, deleteItem } = props;
  const [open, setOpen] = React.useState(false);
  const { setNotification } = useNotification();

  const handleOpen = (): void => {
    setOpen(true);
  };

  const handleClose = (): void => {
    setOpen(false);
  };

  const handleDelete = async (): Promise<void> => {
    const resultAction = await deleteItem({
      categoryId,
      itemId,
    });

    handleClose();

    try {
      unwrapResult(resultAction);
      setNotification({
        severity: "success",
        message: "Delete item successfully",
      });
    } catch (e) {
      setNotification({
        severity: "error",
        message: e.message ?? "Delete item failed",
      });
    }
  };

  return (
    <>
      <Button data-testid="delete-item" color="inherit" onClick={handleOpen}>
        <DeleteIcon />
      </Button>
      <Dialog
        open={open}
        onClose={handleClose}
        aria-labelledby="item-delete-form-title"
        fullWidth
      >
        <DialogTitle id="item-delete-form-title">Delete Item</DialogTitle>
        <DialogContent>
          <DialogContentText>
            Are you sure to delete this item?
          </DialogContentText>
        </DialogContent>
        <DialogActions>
          <Button type="button" onClick={handleClose}>
            Cancel
          </Button>
          <Button color="primary" type="submit" onClick={handleDelete}>
            Delete
          </Button>
        </DialogActions>
      </Dialog>
    </>
  );
};

export default connector(DeleteItem);
