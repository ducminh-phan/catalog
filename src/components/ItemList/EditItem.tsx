import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
} from "@material-ui/core";
import EditIcon from "@material-ui/icons/Edit";
import { unwrapResult } from "@reduxjs/toolkit";
import { TextField } from "mui-rff";
import React, { ReactNode } from "react";
import { Form } from "react-final-form";
import { connect, ConnectedProps } from "react-redux";

import * as itemActions from "actions/item";
import { useNotification } from "contexts/notification";
import * as types from "utils/types";

import { FormData, required, validate } from "./schema";

const mapDispatchToProps = {
  editItem: itemActions.editItem,
};

const connector = connect(null, mapDispatchToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;

type Props = PropsFromRedux & {
  categoryId?: string;
  item: types.Item;
};

const EditItem = (props: Props): React.ReactElement => {
  const { categoryId, editItem, item } = props;
  const [open, setOpen] = React.useState(false);
  const { setNotification } = useNotification();

  const handleOpen = (): void => {
    setOpen(true);
  };

  const handleClose = (): void => {
    setOpen(false);
  };

  const handleEdit = async (data: FormData): Promise<void> => {
    const resultAction = await editItem({
      categoryId,
      itemId: item.id.toString(),
      data,
    });

    handleClose();

    try {
      unwrapResult(resultAction);
      setNotification({
        severity: "success",
        message: "Edit item successfully",
      });
    } catch (e) {
      setNotification({
        severity: "error",
        message: e.message ?? "Edit item failed",
      });
    }
  };

  return (
    <>
      <Button data-testid="edit-item" color="inherit" onClick={handleOpen}>
        <EditIcon />
      </Button>
      <Dialog
        open={open}
        onClose={handleClose}
        aria-labelledby="item-edit-form-title"
        fullWidth
      >
        <DialogTitle id="item-edit-form-title">Edit Item</DialogTitle>
        <DialogContent>
          <Form
            onSubmit={handleEdit}
            initialValues={item}
            validate={validate}
            render={({ handleSubmit, invalid, dirty }): ReactNode => (
              <form onSubmit={handleSubmit} noValidate>
                <TextField
                  id="edit-item-name"
                  label="Name"
                  name="name"
                  required={required.name}
                />
                <TextField
                  id="edit-item-description"
                  label="Description"
                  name="description"
                  required={required.description}
                />
                <TextField
                  id="edit-item-price"
                  label="Price"
                  name="price"
                  type="number"
                  required={required.price}
                />
                <DialogActions>
                  <Button type="button" onClick={handleClose}>
                    Cancel
                  </Button>
                  <Button
                    color="primary"
                    type="submit"
                    disabled={invalid || !dirty}
                  >
                    Submit
                  </Button>
                </DialogActions>
              </form>
            )}
          />
        </DialogContent>
      </Dialog>
    </>
  );
};

export default connector(EditItem);
