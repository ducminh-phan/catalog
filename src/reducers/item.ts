import { createReducer } from "@reduxjs/toolkit";
import { normalize } from "normalizr";

import { editItem, fetchItems } from "actions/item";
import * as schemas from "utils/schemas";
import * as types from "utils/types";

interface ItemState {
  totalItems: number;
  items: {
    [key: string]: types.Item;
  };
  ids: string[];
}

const initialState: ItemState = {
  totalItems: 0,
  items: {},
  ids: [],
};

export default createReducer(initialState, (builder) =>
  builder
    .addCase(fetchItems.fulfilled, (state, action) => {
      const { payload } = action;
      const normalized = normalize(payload.items, schemas.items);

      return {
        totalItems: payload.totalItems,
        items: normalized.entities.items ?? {},
        ids: normalized.result,
      };
    })
    .addCase(editItem.fulfilled, (state, action) => {
      const { payload } = action;
      const normalized = normalize(payload, schemas.item);

      return {
        ...state,
        items: {
          ...state.items,
          ...normalized.entities.items,
        },
      };
    }),
);
