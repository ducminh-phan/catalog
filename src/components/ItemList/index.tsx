import {
  Box,
  List,
  ListItem,
  ListItemSecondaryAction,
  ListItemText,
  Paper,
} from "@material-ui/core";
import Pagination from "@material-ui/lab/Pagination";
import { RouteComponentProps } from "@reach/router";
import React, { ChangeEvent, ReactElement, useEffect, useState } from "react";
import { connect, ConnectedProps } from "react-redux";

import { fetchItems } from "actions/item";
import { useAuth } from "contexts/auth";
import { ITEMS_PER_PAGE } from "enums";
import { RootState } from "reducers";
import * as types from "utils/types";

import EditItem from "./EditItem";

interface ItemListProps extends RouteComponentProps {
  categoryId?: string;
}

const mapDispatchToProps = {
  fetchItems,
};

const connector = connect(
  ({ item }: RootState) => ({ item }),
  mapDispatchToProps,
);

type PropsFromRedux = ConnectedProps<typeof connector>;

type Props = PropsFromRedux & ItemListProps;

const ItemList = (props: Props): ReactElement => {
  // eslint-disable-next-line no-shadow
  const { fetchItems, item, categoryId } = props;

  const { data } = useAuth();
  const [offset, setOffset] = useState(0);
  const [page, setPage] = useState(1);

  useEffect(() => {
    fetchItems({ categoryId, offset, limit: ITEMS_PER_PAGE });
  }, [fetchItems, categoryId, offset]);

  const { totalItems, items } = item;
  const pageCount = Math.ceil(totalItems / ITEMS_PER_PAGE);

  const handleChangePage = (
    event: ChangeEvent<unknown>,
    value: number,
  ): void => {
    setPage(value);
    setOffset(ITEMS_PER_PAGE * (value - 1));
  };

  return (
    <Paper>
      <Box p={2}>
        <List>
          {Object.values(items).map((c: types.Item) => (
            <ListItem key={c.id}>
              <ListItemText primary={c.name} secondary={c.description} />
              {data?.user.id === c.userId ? (
                <ListItemSecondaryAction>
                  <EditItem categoryId={categoryId} item={c} />
                </ListItemSecondaryAction>
              ) : (
                <div />
              )}
            </ListItem>
          ))}
        </List>
        <Pagination count={pageCount} page={page} onChange={handleChangePage} />
      </Box>
    </Paper>
  );
};

export default connector(ItemList);
