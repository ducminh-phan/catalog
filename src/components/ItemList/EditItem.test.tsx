import React from "react";

import store from "store";
import Request from "utils/api";
import {
  fireEvent,
  render,
  setupUser,
  wait,
  waitForElementToBeRemoved,
} from "utils/tests";
import * as types from "utils/types";

import { FetchItemsResponse } from "../../actions/item";
import { act } from "../../utils/tests";
import DeleteItem from "./DeleteItem";
import EditItem from "./EditItem";
import ItemList from "./index";

const categoryId = "0";
const itemId = "1";

it("can edit item", async () => {
  setupUser();

  const items = [1, 2].map((id) => ({
    id,
    userId: 0,
    name: `name ${id}`,
    description: `description ${id}`,
    price: 0,
  }));
  const item = items[parseInt(itemId, 10) - 1];

  jest.spyOn(Request, "get").mockImplementation(
    (): Promise<FetchItemsResponse> =>
      Promise.resolve({
        totalItems: 3,
        items,
      }),
  );

  await act(async () => {
    render(<ItemList categoryId={categoryId} />);
  });
  expect(store.getState().item.totalItems).toBe(3);
  expect(store.getState().item.items).toEqual(
    expect.objectContaining({
      [itemId]: item,
    }),
  );

  const {
    baseElement,
    getByTestId,
    getByText,
    queryByText,
    queryByLabelText,
    getByRole,
    getByLabelText,
  } = render(<EditItem categoryId={categoryId} item={item} />);
  await wait(() => getByTestId("edit-item"));
  expect(baseElement).toMatchSnapshot();

  await wait(() => getByTestId("edit-item"));
  expect(queryByText("Cancel")).toBeNull();

  const editButton = getByTestId("edit-item");

  // Click the Add button
  fireEvent.click(editButton);
  await wait(() => getByText("Cancel"));
  expect(queryByLabelText(/Name.*/)).toBeInTheDocument();
  expect(queryByLabelText(/Description.*/)).toBeInTheDocument();
  expect(queryByLabelText(/Price.*/)).toBeInTheDocument();
  expect(baseElement).toMatchSnapshot();

  // Click the Cancel button
  fireEvent.click(getByText("Cancel"));
  await waitForElementToBeRemoved(() => queryByText("Cancel"));
  expect(queryByLabelText(/Name.*/)).toBeNull();
  expect(queryByLabelText(/Description.*/)).toBeNull();
  expect(queryByLabelText(/Price.*/)).toBeNull();

  // Click the Edit button again
  fireEvent.click(editButton);
  await wait(() => getByText("Cancel"));

  // The submit button should be disabled
  const submitButton = [
    ...getByRole("presentation").querySelectorAll("button"),
  ].filter((node) => node.textContent === "Submit")[0];
  expect(submitButton).toHaveAttribute("disabled");

  // Fill in mock values
  const nameField = getByLabelText(/Name.*/) as HTMLInputElement;
  const mockName = "mockName";
  const mockItem: types.Item = {
    ...item,
    name: mockName,
  };
  await wait(() => {
    fireEvent.change(nameField, {
      target: { value: mockName },
    });
  });

  expect(nameField.value).toBe(mockName);
  expect(submitButton).not.toHaveAttribute("disabled");

  // Submit
  jest
    .spyOn(Request, "put")
    .mockImplementation((): Promise<types.Item> => Promise.resolve(mockItem));

  await wait(() => fireEvent.click(submitButton));
  expect(Request.put).toHaveBeenCalledWith(
    `categories/${categoryId}/items/${itemId}`,
    {
      id: parseInt(itemId, 10),
      name: mockName,
      description: item.description,
      price: item.price,
      userId: 0,
    },
  );

  expect(Object.keys(store.getState().item.items)).toHaveLength(2);
  expect(store.getState().item.items).toEqual(
    expect.objectContaining({ [itemId]: mockItem }),
  );
});
