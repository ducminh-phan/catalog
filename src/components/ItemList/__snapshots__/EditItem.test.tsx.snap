// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can edit item 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiBox-root MuiBox-root"
      >
        <ul
          class="MuiList-root MuiList-padding"
        >
          <li
            class="MuiListItem-root MuiListItem-gutters"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                name 1
              </span>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                description 1
              </p>
            </div>
            <div />
          </li>
          <li
            class="MuiListItem-root MuiListItem-gutters"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                name 2
              </span>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                description 2
              </p>
            </div>
            <div />
          </li>
        </ul>
        <ul
          class="MuiList-root MuiList-padding"
        >
          <li
            class="MuiListItem-container"
          >
            <div
              class="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
            >
              <nav
                aria-label="pagination navigation"
                class="MuiPagination-root"
              >
                <ul
                  class="MuiPagination-ul"
                >
                  <li>
                    <button
                      aria-label="Go to previous page"
                      class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-disabled MuiButtonBase-disabled"
                      disabled=""
                      tabindex="-1"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiPaginationItem-icon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                        />
                      </svg>
                    </button>
                  </li>
                  <li>
                    <button
                      aria-current="true"
                      aria-label="page 1"
                      class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-selected"
                      tabindex="0"
                      type="button"
                    >
                      1
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </button>
                  </li>
                  <li>
                    <button
                      aria-label="Go to page 2"
                      class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
                      tabindex="0"
                      type="button"
                    >
                      2
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </button>
                  </li>
                  <li>
                    <button
                      aria-label="Go to next page"
                      class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
                      tabindex="0"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiPaginationItem-icon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </button>
                  </li>
                </ul>
              </nav>
            </div>
            <div
              class="MuiListItemSecondaryAction-root"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
                data-testid="add-item"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
      data-testid="edit-item"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</body>
`;

exports[`can edit item 2`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiBox-root MuiBox-root"
      >
        <ul
          class="MuiList-root MuiList-padding"
        >
          <li
            class="MuiListItem-root MuiListItem-gutters"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                name 1
              </span>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                description 1
              </p>
            </div>
            <div />
          </li>
          <li
            class="MuiListItem-root MuiListItem-gutters"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                name 2
              </span>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                description 2
              </p>
            </div>
            <div />
          </li>
        </ul>
        <ul
          class="MuiList-root MuiList-padding"
        >
          <li
            class="MuiListItem-container"
          >
            <div
              class="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
            >
              <nav
                aria-label="pagination navigation"
                class="MuiPagination-root"
              >
                <ul
                  class="MuiPagination-ul"
                >
                  <li>
                    <button
                      aria-label="Go to previous page"
                      class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-disabled MuiButtonBase-disabled"
                      disabled=""
                      tabindex="-1"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiPaginationItem-icon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                        />
                      </svg>
                    </button>
                  </li>
                  <li>
                    <button
                      aria-current="true"
                      aria-label="page 1"
                      class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-selected"
                      tabindex="0"
                      type="button"
                    >
                      1
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </button>
                  </li>
                  <li>
                    <button
                      aria-label="Go to page 2"
                      class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
                      tabindex="0"
                      type="button"
                    >
                      2
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </button>
                  </li>
                  <li>
                    <button
                      aria-label="Go to next page"
                      class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
                      tabindex="0"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiPaginationItem-icon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root"
                      />
                    </button>
                  </li>
                </ul>
              </nav>
            </div>
            <div
              class="MuiListItemSecondaryAction-root"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
                data-testid="add-item"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
  <div
    aria-hidden="true"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
      data-testid="edit-item"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    class="MuiDialog-root"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper"
      role="none presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-labelledby="item-edit-form-title"
        class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiDialog-paperFullWidth MuiPaper-elevation24 MuiPaper-rounded"
        role="dialog"
      >
        <div
          class="MuiDialogTitle-root"
          id="item-edit-form-title"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6"
          >
            Edit Item
          </h2>
        </div>
        <div
          class="MuiDialogContent-root"
        >
          <form
            novalidate=""
          >
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled MuiFormLabel-required MuiInputLabel-required"
                data-shrink="true"
                for="edit-item-name"
                id="edit-item-name-label"
              >
                Name
                <span
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input"
                  id="edit-item-name"
                  name="name"
                  required=""
                  type="text"
                  value="name 1"
                />
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled MuiFormLabel-required MuiInputLabel-required"
                data-shrink="true"
                for="edit-item-description"
                id="edit-item-description-label"
              >
                Description
                <span
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input"
                  id="edit-item-description"
                  name="description"
                  required=""
                  type="text"
                  value="description 1"
                />
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled MuiFormLabel-required MuiInputLabel-required"
                data-shrink="true"
                for="edit-item-price"
                id="edit-item-price-label"
              >
                Price
                <span
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input"
                  id="edit-item-price"
                  name="price"
                  required=""
                  type="number"
                  value="0"
                />
              </div>
            </div>
            <div
              class="MuiDialogActions-root MuiDialogActions-spacing"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  Cancel
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-disabled MuiButtonBase-disabled"
                disabled=""
                tabindex="-1"
                type="submit"
              >
                <span
                  class="MuiButton-label"
                >
                  Submit
                </span>
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
